src/vs/code/electron-main/app.ts

```diff
@@ -846,6 +851,68 @@ export class CodeApplication extends Disposable {
 	}
 }
 
+class WebviewProtocolProvider extends Disposable {
+
+	private readonly webviewMetadata = new Map<string, {
+		readonly extensionLocation: URI | undefined;
+		readonly localResourceRoots: URI[];
+	}>();
+
+	constructor(
+		@IFileService private readonly fileService: IFileService,
+	) {
+		super();
+
+		ipc.on('vscode:registerWebview', (event: IpcMainEvent, id: string, data: any) => {
+			if (!this.webviewMetadata.has(id)) {
+				const ses = session.fromPartition(id);
+				this.register(ses);
+			}
+
+			this.webviewMetadata.set(id, {
+				extensionLocation: data.extensionLocation ? URI.from(data.extensionLocation) : undefined,
+				localResourceRoots: Array.isArray(data.localResourceRoots) ? data.localResourceRoots.map((x: UriComponents) => URI.from(x)) : []
+			});
+
+			event.sender.send(`vscode:didRegisterWebview-${id}`);
+		});
+
+		ipc.on('vscode:unregisterWebview', (_event: IpcMainEvent, id: string) => {
+			this.webviewMetadata.delete(id);
+		});
+	}
+
+	private register(session: Session) {
+		session.protocol.registerBufferProtocol(Schemas.vscodeWebviewResource, async (request, callback: any) => {
```

Comment:
I would strongly recommend to work with streams and not buffer, there seems to be `registerStreamProtocol` for that purpose?